#8003 Adding "Add a new micro-ROS app" (#125)
micro-ROS/micro-ROS.github.io: _docs/tutorials/advanced/nuttx/add_a_microros_application/index.md

---
title: Add a new Micro-ROS application
permalink: /docs/tutorials/advanced/nuttx/add_a_microros_application/
---
# Add a Micro-ROS application to NuttX

In this tutorial, we will show the process of how to add a new micro-ROS application to NuttX.
This application is a basic example of a Ping-Pong Micro-ROS application over Ethernet communication.

> ## Build the Micro-ROS build system

Reviewer: maybe here before the box, a line explaining what you do (source the opportune ROS version and build the micro-ROS packages)

Author: I've added a table with the details of the rtos, board and ros2 version.

```bash
source /opt/ros/$ROS_DISTRO/setup.bash

mkdir uros_ws && cd uros_ws

git clone -b $ROS_DISTRO https://github.com/micro-ROS/micro-ros-build.git src/micro-ros-build

rosdep update && rosdep install --from-path src --ignore-src -y

colcon build

source install/local_setup.bash
```

> ## Create micro-ROS firmware

Reviewer: Maybe at some point explain that this tutorial works for NuttX on the OLIMEX STM32 etc board

This step will download all the required software.
```bash
ros2 run micro_ros_setup create_firmware_ws.sh nuttx olimex-stm32-e407
```

> At this moment we have everything ready in our workspace to develop a new application. On the next steps, we will start the development of the app itself.elf.

Reviewer: ```suggestion
At this point we have everything ready in our workspace to develop a new application. With the next steps, we will start the development of the app itself.
```

## Aplication development

- Go to: ``uros_ws/firmware/apps/example``
- Create a folder called ``uros_pingpong``

### Create Kconfig file

> On this step, you need to create an empty file called **Kconfig**, inside the application folder. These files contain the configuration menu to configure the application and add to the build system list of NuttX.

Reviewer: ```suggestion
For this step, you need to create an empty file called **Kconfig**, inside the application folder. These files contain the configuration menu to configure the application and add it to the build system list of NuttX.
```

> On the next lines you can see an example of a basic generic implementation:

Reviewer: ```suggestion
In the next lines you can see an example of a basic generic implementation:
```

```bash
config UROS_PINGPONG_EXAMPLE
	bool "micro-ROS Ping Pong"
	default n
	depends on UROS
	---help---
		<Brief description of the application>

if UROS_PINGPONG_EXAMPLE

config UROS_PINGPONG_EXAMPLE_PROGNAME
	string "Program name"
	default "uros_ping_pong"
	depends on BUILD_KERNEL
	---help---
		This is the name of the program that will be use when the NSH ELF
		program is installed.

endif
```

### Create Makefile 

Create an empty file called **Makefile**, inside of the application folder.

> On the next lines you can see an example of a generic implementation:

Reviewer: ```suggestion
Below, you can see an example of a generic implementation:
```

```Makefile
-include $(TOPDIR)/Make.defs

#Set the scheduler priority for the app.
CONFIG_UROS_PINGPONG_EXAMPLE_PRIORITY ?= SCHED_PRIORITY_DEFAULT
#Set the stack size to the app. The minimum stack size on NuttX for a micro-ROS App is 65000 bytes
CONFIG_UROS_PINGPONG_EXAMPLE_STACKSIZE ?= 65000

#This is the name of the app on the NSH console
APPNAME = uros_pingpong
PRIORITY = $(CONFIG_UROS_PINGPONG_EXAMPLE_PRIORITY)
STACKSIZE = $(CONFIG_UROS_PINGPONG_EXAMPLE_STACKSIZE)

#Add the source files.
ASRCS =
CSRCS =
MAINSRC = uros_pingpong_main.c 

CONFIG_UROS_PINGPONG_EXAMPLE_PROGNAME ?= uros_pingpong$(EXEEXT)
PROGNAME = $(CONFIG_UROS_PINGPONG_EXAMPLE_PROGNAME)
UROS_PINGPONG_INCLUDES = $(shell find $(APPDIR)/$(CONFIG_UROS_DIR)/install -type d -name include)
CFLAGS += ${shell $(INCDIR) $(INCDIROPT) "$(CC)" "$(UROS_PINGPONG_INCLUDES)"} -std=c99

MODULE = CONFIG_UROS_PINGPONG_EXAMPLE

include $(APPDIR)/Application.mk
```

### Create Make.defs file
Create an empty file called **Make.defs**, inside of the application folder.

> You need to add the next lines:

Reviewer: ```suggestion
You need to add the following lines:
```

```Makefile
ifeq ($(CONFIG_UROS_PINGPONG_EXAMPLE),y)
    CONFIGURED_APPS += examples/uros_pingpong
endif
```

This file will add the folder as an application if you set the option on the configuration menu of NuttX. 

### Create the main file

Create a file named ``uros_pingpong_main.c``.

> On the next lines we will show a pinpong example:

Reviewer: ```suggestion
In the box below, we will show a pinpong example:
```

      msg.stamp.sec = ts.tv_sec;
      msg.stamp.nanosec = ts.tv_nsec;

      // Reset the pong count and publish the ping message
      pong_count = 0;
      rcl_publish(&ping_publisher, (const void*)&msg, NULL);
      // printf("Ping send seq 0x%x\n", seq_no);
    }
    
    // Check if some pong message is received
    if (wait_set.subscriptions[index_pong_subscription]) {
      rc = rcl_take(wait_set.subscriptions[index_pong_subscription], &rcv_msg, NULL, NULL);

      if(rc == RCL_RET_OK && strcmp(msg.frame_id.data,rcv_msg.frame_id.data) == 0) {
          pong_count++;
          // printf("Pong for seq 0x%x (%d)\n", seq_no, pong_count);
      }
    }

    // Check if some ping message is received and pong it
    if (wait_set.subscriptions[index_ping_subscription]) {
      rc = rcl_take(wait_set.subscriptions[index_ping_subscription], &rcv_msg, NULL, NULL);

      // Dont pong my own pings
      if(rc == RCL_RET_OK && strcmp(msg.frame_id.data,rcv_msg.frame_id.data) != 0){
        // printf("Ping received with seq 0x%x (%d). Answering.\n", seq_no);
        rcl_publish(&pong_publisher, (const void*)&rcv_msg, NULL);
      }
    }
    
    usleep(10000);
  } while (true);
}
```

> Now the application is ready, it only left to create the right configuration.

Reviewer: ```suggestion
Once the application is ready, the only thing left to do is to create the right configuration.
```

## Create a config profile

> Create a specific configuration from scratch can be a challenging task, so we're going to start from an already existing one and modify it to our new application.

Reviewer: ```suggestion
Create a specific configuration from scratch can be a challenging task, so we're going to start from an already existing one and modify it for our new application.
```

Execute the next command:
```bash
ros2 run micro_ros_setup configure_firmware.sh uros
```

> This set the Ethernet and Micro-ROS required configuration, but we're going to modify it, to add our application.

Reviewer: ```suggestion
This sets the Ethernet and Micro-ROS required configuration. However, in order to add our application, we're going to modify it.
```

### Modify the configuration
On a console, type the next commands:
```bash
cd uros_ws/firmware/nuttx
make menuconfig
```

> This will open the NuttX menu config, which allows you to modify the configuration of the RTOS, including add a new application.

Reviewer: ```suggestion
This will open the NuttX menu config, which allows you to modify the configuration of the RTOS, including adding a new application.
```

<!-- Add image of the menu-->

- On the menu, you need to follow the next path:
``Application Configuration -> Examples ``
![](images/nuttx_menuconfig.png)

> - This will show a list of the available application, you need to find: ``micro-ROS Ping-Pong`` and click ``y`` to add.

Reviewer: ```suggestion
- This will show a list of the available applications. You need to find: ``micro-ROS Ping-Pong`` and click ``y`` to add it.
```

![](images/nuttx_examples.png)

> - Now push three times the key ``ESC`` to close the menu. This will ask you if you want to save the save your new configuration, and you need to push ``Yes``

Reviewer: ```suggestion
- Now push three times the key ``ESC`` to close the menu. You will be asked if you want to save the save your new configuration, and you need to click ``Yes``.
```

<!-- Add image of the save screen-->

### Add your configuration

> Open a console and execute the next commands:

Reviewer: ```suggestion
Open a console and execute the following commands:
```

```bash
cd uros_ws/firmware/nuttx
make savedefconfig
```

This will generate a file called ``defconfig`` on ``uros_ws/firmware/nuttx``. This file is a config profile with all the required configuration to run your specific configuration which includes your application.

Finally create a folder called ``uros_pingpong`` on ``uros_ws/firmware/nuttx/configs/olimex-stm32-e407`` and move the defconfig file to uros_pingpong folder.

## Build your application

> Now is everything ready it only left to continue with the build system instructions. Execute the next command on the build system to start the build process:

Reviewer: ```suggestion
Now that everything is ready, the only thing left to do is to continue with the build system instructions. Execute the next command on the build system to start the build process:
```

```bash
ros2 run micro_ros_setup configure_firmware.sh uros_pingpong
ros2 run micro_ros_setup build_firmware.sh
```

> If it returns the next output, your build succed and your application was implemented properly.

Reviewer: ```suggestion
If it returns the next output, your build step succeded and your application was implemented properly.
```

| NuttX | Olimex-STM32-E407 |    Dashing   |

In this tutorial, we will follow a guide of how to **add a new Micro-ROS application** on the NuttX RTOS. During this guide, we will start with how to create a configuration profile for our application, following of how to create a micro-ROS pingpong application and finalizing of how we can add our new application to NuttX.

## Build the Micro-ROS build system

```bash
source /opt/ros/$ROS_DISTRO/setup.bash

mkdir uros_ws && cd uros_ws

git clone -b $ROS_DISTRO https://github.com/micro-ROS/micro-ros-build.git src/micro-ros-build

rosdep update && rosdep install --from-path src --ignore-src -y

colcon build

source install/local_setup.bash
```
## Create micro-ROS firmware

This step will download all the required software.
```bash
ros2 run micro_ros_setup create_firmware_ws.sh nuttx olimex-stm32-e407
```
At this point we have everything ready in our workspace to develop a new application. With the next steps, we will start the development of the app itself.

## Aplication development

- Go to: ``uros_ws/firmware/apps/example``
- Create a folder called ``uros_pingpong``

### Create Kconfig file

> For this step, you need to create an empty file called **Kconfig**, inside the application folder. These files contain the configuration menu to configure the application and add it to the build system list of NuttX.

Author: Added previous explanation of which is the application folder.

    if (wait_set.subscriptions[index_pong_subscription]) {
      rc = rcl_take(wait_set.subscriptions[index_pong_subscription], &rcv_msg, NULL, NULL);

      if(rc == RCL_RET_OK && strcmp(msg.frame_id.data,rcv_msg.frame_id.data) == 0) {
          pong_count++;
          // printf("Pong for seq 0x%x (%d)\n", seq_no, pong_count);
      }
    }

    // Check if some ping message is received and pong it
    if (wait_set.subscriptions[index_ping_subscription]) {
      rc = rcl_take(wait_set.subscriptions[index_ping_subscription], &rcv_msg, NULL, NULL);

      // Dont pong my own pings
      if(rc == RCL_RET_OK && strcmp(msg.frame_id.data,rcv_msg.frame_id.data) != 0){
        // printf("Ping received with seq 0x%x (%d). Answering.\n", seq_no);
        rcl_publish(&pong_publisher, (const void*)&rcv_msg, NULL);
      }
    }
    
    usleep(10000);
  } while (true);
}
```

Once the application is ready, the only thing left to do is to create the right configuration.

## Create a config profile

Create a specific configuration from scratch can be a challenging task, so we're going to start from an already existing one and modify it for our new application.

> Execute the next command:

Author: Modify commands to clarify where you need to execute the commands

---
title: Add a new Micro-ROS application
permalink: /docs/tutorials/advanced/nuttx/add_a_microros_application/
---

|  RTOS | Board Compatible  | ROS2 Version |
|:-----:|-------------------|:------------:|
| NuttX | Olimex-STM32-E407 |    Dashing   |

In this tutorial, we will follow a guide of how to **add a new Micro-ROS application** on the NuttX RTOS. During this guide, we will start with how to create a configuration profile for our application, following of how to create a micro-ROS pingpong application and finalizing of how we can add our new application to NuttX.

## Build the Micro-ROS build system

```bash
source /opt/ros/$ROS_DISTRO/setup.bash

mkdir uros_ws && cd uros_ws

git clone -b $ROS_DISTRO https://github.com/micro-ROS/micro-ros-build.git src/micro-ros-build

> rosdep update && rosdep install --from-path src --ignore-src -y

Reviewer 2: ```suggestion
sudo apt update && rosdep update
rosdep install --from-path src --ignore-src -y
```
